Move failed and rejected CFDs to a separate DB table

luckysori opened the pull request:
Fixes #1881.

Tasks:

- [x] Refactor `db` so that it doesn't keep growing out of control.
- [x] Add `db` APIs to move failed and rejected CFDs to a separate table.
- [x] Test `db::failed` APIs.
- [x] Introduce actor to periodically move failed and rejected CFDs to the corresponding table.
- [x] Consider removing any filters applied to the `cfds` table (since it can now be assumed that it generally only contains open CFDs).

luckysori committed:
Remove unnecessary Unpin trait bound

luckysori committed:
Make trait bound of load_all_cfds explicit

It is technically not necessary because `ClosedCfdAggregate` implies
`CfdAggregate`, but it makes the code easier to understand.

luckysori committed:
Log on success when moving CFDs to closed_cfds table

thomaseizinger commented on daemon/src/db/failed.rs:
```diff
@@ -0,0 +1,179 @@
+//! This module allows us to move faile CFDs to a separate table so
```
```suggestion
//! This module allows us to move failed CFDs to a separate table so
```

thomaseizinger commented on daemon/src/db/failed.rs:
```diff
@@ -0,0 +1,179 @@
+                tracing::warn!(order_id = %id, "Failed to move failed CFD: {e:#}");
+            }
+
+            tracing::debug!(order_id = %id, "Moved CFD to `failed_cfds` table");
```
You will print this even in an error case :)

The entire construct should probably be a `match`!

thomaseizinger commented on daemon/src/db/closed.rs:
```diff
@@ -142,6 +142,8 @@ impl Connection {
                 tracing::warn!(order_id = %id, "Failed to move closed CFD: {e:#}");
             }
+
+            tracing::debug!(order_id = %id, "Moved CFD to `closed_cfds` table");
```
This will log-unconditionally, so also in the case of an error above :)

Probably best to use a `match` or early return.

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -315,6 +317,28 @@ impl Connection {
+                    failed_cfds
+                "#
+            )
+            .fetch_all(&mut *conn)
```
I recently learned that there is also `fetch` which returns a `Stream`! That would make the whole thing even more "iteratively"!

luckysori replied:
Thanks for the tip!

How do you think it would interact with the fact that we would be using the same `Connection` to generate this stream and to load CFDs?

thomaseizinger replied:
Yeah you would have to use the same connection likely. One thing that you might be able to do is do the mapping right on the stream that is being returned instead of collecting it all first. Doesn't make much of a difference though I think!

luckysori replied:
In the interest of getting this merged before I encounter more conflicts I vote for merging this as is and I can look at this suggestion in a follow-up PR.

thomaseizinger reviewed (commented):
Few comments but looks like a good direction!

da-kami commented on daemon/src/db/failed.rs:
```diff
@@ -0,0 +1,179 @@
+
+/// Data loaded from the database about a failed CFD.
+#[derive(Debug, Clone, Copy)]
+pub struct FailedCfd {
```
🛠 Failed CFDs may have events, i.e. we may have a `ContractSetupStarted` and `ContractSetupFailed` event. I think it might be a good idea to save the timestamp of the first event recorded - that is what we use for sorting the open CFDs.. 
I think a user would still expect to see the CFDs chronologically ordered, but without timestamps we cannot really achieve that. 
I think this is actually already missing when processing closed cfds - I would always save it as `creation_timestamp` (or `first_event_timestamp` it you want). I am not sure we have to add it as another field, we could also load it from the eventlog (I suppose failed will also have eventlog).
I think we should always have at least one event in the database, so we should always be able to come up with a timestamp. For cases where we don't have any events we could actually just delete from `cfds` because this is too incomplete to do anything.

luckysori replied:
> 🛠 Failed CFDs may have events, i.e. we may have a ContractSetupStarted and ContractSetupFailed event.

Do you think it's worth saving events for failed CFDs? To me it's even questionable if we should be persisting them! One could imagine that a failed CFD could eventually just become an ephemeral notification for the user.

But with this PR I would persist them since we have no better alternative at the moment.

> I think it might be a good idea to save the timestamp of the first event recorded - that is what we use for sorting the open CFDs..
I think a user would still expect to see the CFDs chronologically ordered, but without timestamps we cannot really achieve that.

Good point: we want these to appear in order. It is a bit annoying to have to process the events at all just to get a timestamp, but I guess we should.

> I think this is actually already missing when processing closed cfds - I would always save it as creation_timestamp (or first_event_timestamp it you want).

It's true that the closed CFDs are not sorted properly (because we don't use the `projection::Cfd::aggregate` field. More importantly, the `profit_percent`, `profit_btc` and `payout` fields are actually wrong. I added a test to verify this, but I did not realise that we were calling `with_current_quote` on all CFDs (doing so on closed CFDs overwrites these values with wrong ones).

> I am not sure we have to add it as another field, we could also load it from the eventlog (I suppose failed will also have eventlog).

I would prefer not to add entries to the `event_log` table for failed or rejected CFDs, because they're always going to be the same. I think this information wouldn't be very useful.

da-kami replied:
> I would prefer not to add entries to the event_log table for failed or rejected CFDs, because they're always going to be the same. I think this information wouldn't be very useful.

I am not sure about that. In *most* cases it won't have value, but in some cases it might. There might be error scenarios where something goes into a weird failed state that should be analysed. It's not expensive to save these event log entries. 
If we decide to not keep an eventlog for failed at all we *definitely* need a timestamp in the table, because otherwise we don't know anything about the time the CFD "failed". I would say it might actually be valuable to know when contract setup started and when it failed. (Rejected might be interesting as well, because we have reasons in there - but that does not necessarily have to be stored in an event-log; the reason would not be persisted there...).

da-kami reviewed (commented):
Overall concept ack. 
Would you treat rejected separate or handle them as "failed" where we add a reason why it failed?
I think it might be good enough to treat all "failed" in the same table unless we want to store different data for failed/rejected.

I noticed that we are missing one thing from the previous work already and that is the ability to sort the closed / failed CFDs chronologically. See my related comment

luckysori committed:
Load creation_timestamp for closed CFDs

This ensures that closed CFDs can actually be sorted properly in
`projection`.

luckysori committed:
Extract database APIs related to closed CFDs into submodule

luckysori committed:
Archive failed CFDs every half hour

luckysori committed:
Change terminology around close_cfds actor

To match the new terminology introduced with the
`archive_failed_cfds::Actor`.

luckysori committed:
Introduce db APIs to move failed CFDs to separate table

luckysori committed:
Simplify query for loading all open CFD IDs

Since we now periodically move CFDs that are failed and closed to
separate tables, we don't really need to run a relatively expensive
query every time we want to get open CFDs.

This does mean that sometimes we may return CFDs that are no longer
relevant, but we are willing to accept these transient inconsistencies
in the name of performance.

luckysori committed:
Extract function to load all closed CFD IDs

luckysori committed:
Extract function to load all failed CFD IDs

da-kami commented on daemon/src/db/failed.rs:
```diff
@@ -0,0 +1,556 @@
+        let ids = self.failed_cfd_ids_according_to_events().await?;
+
+        if !ids.is_empty() {
+            tracing::debug!("Moving ContractSetupFailed CFDs to failed_cfds table: {ids:?}");
```
```suggestion
            tracing::debug!("Moving failed CFDs to failed_cfds table: {ids:?}");
```

It includes both `ContractSetupFailed` and `OfferRejected`

da-kami commented on daemon/src/db/failed.rs:
```diff
@@ -0,0 +1,556 @@
+    pub n_contracts: Contracts,
+    pub counterparty_network_identity: Identity,
+    pub role: Role,
+    pub fees: Fees,
```
I remember we discussed this: The `fees` are added because we display them in the UI?
Maybe we should actually re-consider this. It is somewhat confusing and should always be `0`. I think the UI is also confusing for displaying it.

We don't have to block this PR by this change, can do in a follow-up.

luckysori replied:
Fair point! I just wanted to keep the behaviour the same as it currently is in production for the time being.

da-kami commented on daemon/src/db/failed.rs:
```diff
@@ -0,0 +1,556 @@
+        JOIN
+            failed_cfds on failed_cfds.id = event_log_failed.cfd_id
+        WHERE
+            failed_cfds.uuid = $1 AND (event_log_failed.name = $2 OR event_log_failed.name = $3)
```
🛠 I read the comment, but I am not sure why we cannot always `select` the first event in the log. I would prefer that over the current solution. I think an `ASC LIMIT 1` should do the job or am I missing something?

luckysori replied:
Fair enough, I did not consider taking the first event! I'll try that instead.

da-kami commented on daemon/src/projection.rs:
```diff
@@ -971,7 +974,83 @@ impl db::ClosedCfdAggregate for Cfd {
+        // which is why this field is mostly ignored
+        let mut aggregated = Aggregated::new(FeeAccount::new(position, role));
+
+        // set the creation_timestamp to be able to sort closed CFDs
```
```suggestion
        // set the creation_timestamp to be able to sort failed CFDs
```

da-kami commented on daemon/src/projection.rs:
```diff
@@ -971,7 +974,83 @@ impl db::ClosedCfdAggregate for Cfd {
+            Position::Short => calculate_short_liquidation_price(our_leverage, initial_price),
+        };
+
+        // there are no events to apply at this stage for closed CFDs,
```
```suggestion
        // there are no events to apply at this stage for failed CFDs,
```

da-kami commented on daemon/src/projection.rs:
```diff
@@ -971,7 +974,83 @@ impl db::ClosedCfdAggregate for Cfd {
+    }
+}
+
+impl db::FailedCfdAggregate for Cfd {
```
❓Can we potentially create a generic aggregate for closed and failed?

luckysori replied:
I don't see it at the moment because we have completely different expectations for closed and failed CFDs. The data that is stored is different, particularly in terms of the transactions that are attached to closed CFDs (always a lock transaction and a choice of settlement transactions) and that are not attached to failed CFDs. 

We could define a model with optional fields, but I'm not sure I would recommend that added complexity.

da-kami reviewed (approved):
Thanks for the great work! 
Nothing blocking per-se, but I think selecting the `creation_timestamp` for failed cfds could be changed.

da-kami commented on daemon/src/db/closed.rs:
```diff
@@ -0,0 +1,1365 @@
+/// Every closed CFD must have gone through contract setup at some
+/// point. Therefore, we base the creation timestamp on the
+/// `EventKind::ContractSetupStarted` variant.
+async fn load_creation_timestamp(
```
We could actually use the logic that is used for `failed_cfds` here as well. Can be done in follow-up.

luckysori replied:
Done with https://github.com/itchysats/itchysats/pull/1916/commits/38c75bf4010bccb508edf0a6e0da8ead1c93409e.

luckysori committed:
Simplify query to load creation timestamp of closed CFDs

luckysori committed:
Differentiate between refunded and closed in position metrics

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -431,6 +455,37 @@ impl Connection {
+        .await?
+        .into_iter()
+        .map(|r| r.uuid)
+        .collect();
```
Here we could omit the `into_iter` part of we would use `fetch`.

luckysori replied:
I don't seem to be able to return the stream because `conn` is borrowed and dropped as soon as we return :(

```rust
// error[E0515]: cannot return value referencing local variable `conn`
//    --> daemon/src/db.rs:391:9
//     |
// 388 |         .fetch(&mut *conn)
//     |                      ---- `conn` is borrowed here
// ...
// 391 |         Ok(ids)
//     |         ^^^^^^^ returns a value referencing data owned by the current function

async fn closed_cfd_ids_according_to_the_blockchain(
    &self,
) -> Result<impl Stream<Item = Result<OrderId, sqlx::Error>>> {
    let mut conn = self.inner.acquire().await?;

    let ids = sqlx::query!(
        r#"
        select
            id as cfd_id,
            uuid as "uuid: model::OrderId"
        from
            cfds
        where exists (
            select id from EVENTS as events
            where events.cfd_id = cfds.id and
            (
                events.name = $1 or
                events.name = $2 or
                events.name= $3
            )
        )
        "#,
        EventKind::COLLABORATIVE_SETTLEMENT_CONFIRMED,
        EventKind::CET_CONFIRMED,
        EventKind::REFUND_CONFIRMED,
    )
    .fetch(&mut *conn)
    .map(|res| res.map(|r| r.uuid));

    Ok(ids)
}

```

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -264,23 +264,7 @@ impl Connection {
-
-            drop(conn);
-
+            let ids = self.load_open_cfd_ids().await?;
```
We should probably document that callers cannot rely on only open IDs being returned here.

thomaseizinger reviewed (approved):
Excellent work @luckysori !
